File: mt-pricing-frontend/lib/api/endpoints/procurement.ts
"use client";

import env from "@/lib/env";
import { createSupabaseBrowserClient } from "@/lib/supabase/client";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type PRStatus =
  | "draft"
  | "pending_approval"
  | "approved"
  | "rejected"
  | "cancelled"
  | "converted_to_po";

export interface PurchaseRequisitionRead {
  id: string;
  pr_number: string;
  requester_id: string;
  product_sku: string | null;
  qty: string;
  uom: string;
  required_date: string | null;
  cost_center_id: string | null;
  suggested_vendor_id: string | null;
  estimated_amount: string | null;
  status: PRStatus;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface PRCreatePayload {
  product_sku?: string | null;
  qty: string;
  uom?: string;
  required_date?: string | null;
  cost_center_id?: string | null;
  suggested_vendor_id?: string | null;
  estimated_amount?: string | null;
  notes?: string | null;
}

export interface PRFilters {
  status?: PRStatus | "" | undefined;
  cursor?: string | null | undefined;
  limit?: number | undefined;
}

export interface ApprovalRuleRead {
  id: string;
  document_type: string;
  min_amount: string;
  max_amount: string | null;
  category_id: string | null;
  approver_role: string | null;
  approver_user_id: string | null;
  timeout_hours: number;
  priority: number;
  is_active: boolean;
  created_at: string;
}

export interface ApprovalRuleCreatePayload {
  document_type?: string;
  min_amount?: string;
  max_amount?: string | null;
  category_id?: string | null;
  approver_role?: string | null;
  approver_user_id?: string | null;
  timeout_hours?: number;
  priority?: number;
  is_active?: boolean;
}

export interface VendorConditionRead {
  id: string;
  vendor_id: string;
  product_sku: string;
  price: string;
  uom: string;
  moq: number;
  lead_time_days: number | null;
  valid_from: string;
  valid_to: string | null;
  currency: string;
  is_active: boolean;
  created_at: string;
}

export interface VendorConditionCreatePayload {
  vendor_id: string;
  product_sku: string;
  price: string;
  uom?: string;
  moq?: number;
  lead_time_days?: number | null;
  valid_from?: string | null;
  valid_to?: string | null;
  currency?: string;
  is_active?: boolean;
}

export interface VendorConditionUpdatePayload {
  price?: string;
  uom?: string;
  moq?: number;
  lead_time_days?: number | null;
  valid_to?: string | null;
  currency?: string;
  is_active?: boolean;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

export class ProcurementApiError extends Error {
  public readonly status: number;
  public readonly detail: unknown;
  constructor(status: number, detail: unknown, fallback: string) {
    super(typeof detail === "string" ? detail : fallback);
    this.name = "ProcurementApiError";
    this.status = status;
    this.detail = detail;
  }
}

async function authedFetch<T>(path: string, init: RequestInit = {}): Promise<T> {
  const supabase = createSupabaseBrowserClient();
  const {
    data: { session },
  } = await supabase.auth.getSession();
  const headers = new Headers(init.headers);
  if (!headers.has("Content-Type") && init.body) {
    headers.set("Content-Type", "application/json");
  }
  if (session?.access_token) {
    headers.set("Authorization", `Bearer ${session.access_token}`);
  }
  const res = await fetch(`${env.NEXT_PUBLIC_BACKEND_URL}${path}`, {
    ...init,
    headers,
    cache: "no-store",
  });
  if (!res.ok) {
    let detail: unknown;
    try {
      detail = await res.json();
    } catch {
      /* noop */
    }
    throw new ProcurementApiError(res.status, detail, res.statusText);
  }
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

function buildQuery(
  params: Record<string, string | number | boolean | undefined | null>,
): string {
  const search = new URLSearchParams();
  Object.entries(params).forEach(([k, v]) => {
    if (v === undefined || v === null || v === "") return;
    search.set(k, String(v));
  });
  const s = search.toString();
  return s ? `?${s}` : "";
}

const BASE = "/api/v1/procurement";

// ---------------------------------------------------------------------------
// API
// ---------------------------------------------------------------------------

export const procurementApi = {
  // --- Purchase Requisitions -----------------------------------------------
  listRequisitions: (filters: PRFilters = {}): Promise<PurchaseRequisitionRead[]> =>
    authedFetch<PurchaseRequisitionRead[]>(
      `${BASE}/requisitions${buildQuery({
        status: filters.status,
        cursor: filters.cursor ?? undefined,
        limit: filters.limit,
      })}`,
    ),

  getRequisition: (id: string): Promise<PurchaseRequisitionRead> =>
    authedFetch<PurchaseRequisitionRead>(`${BASE}/requisitions/${id}`),

  createRequisition: (payload: PRCreatePayload): Promise<PurchaseRequisitionRead> =>
    authedFetch<PurchaseRequisitionRead>(`${BASE}/requisitions`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  submitRequisition: (id: string): Promise<PurchaseRequisitionRead> =>
    authedFetch<PurchaseRequisitionRead>(`${BASE}/requisitions/${id}/submit`, {
      method: "PATCH",
    }),

  approveRequisition: (id: string): Promise<PurchaseRequisitionRead> =>
    authedFetch<PurchaseRequisitionRead>(`${BASE}/requisitions/${id}/approve`, {
      method: "PATCH",
    }),

  rejectRequisition: (id: string, reason: string): Promise<PurchaseRequisitionRead> =>
    authedFetch<PurchaseRequisitionRead>(`${BASE}/requisitions/${id}/reject`, {
      method: "PATCH",
      body: JSON.stringify({ reason }),
    }),

  cancelRequisition: (id: string): Promise<PurchaseRequisitionRead> =>
    authedFetch<PurchaseRequisitionRead>(`${BASE}/requisitions/${id}/cancel`, {
      method: "PATCH",
    }),

  // --- Approval Rules -------------------------------------------------------
  listApprovalRules: (): Promise<ApprovalRuleRead[]> =>
    authedFetch<ApprovalRuleRead[]>(`${BASE}/approval-rules`),

  createApprovalRule: (payload: ApprovalRuleCreatePayload): Promise<ApprovalRuleRead> =>
    authedFetch<ApprovalRuleRead>(`${BASE}/approval-rules`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  updateApprovalRule: (
    id: string,
    payload: Partial<ApprovalRuleCreatePayload>,
  ): Promise<ApprovalRuleRead> =>
    authedFetch<ApprovalRuleRead>(`${BASE}/approval-rules/${id}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    }),

  // --- Vendor Conditions (PIR) ----------------------------------------------
  listVendorConditions: (params: {
    vendor_id?: string;
    product_sku?: string;
    active_only?: boolean;
  } = {}): Promise<VendorConditionRead[]> =>
    authedFetch<VendorConditionRead[]>(
      `${BASE}/vendor-conditions${buildQuery(params)}`,
    ),

  createVendorCondition: (
    payload: VendorConditionCreatePayload,
  ): Promise<VendorConditionRead> =>
    authedFetch<VendorConditionRead>(`${BASE}/vendor-conditions`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  updateVendorCondition: (
    id: string,
    payload: VendorConditionUpdatePayload,
  ): Promise<VendorConditionRead> =>
    authedFetch<VendorConditionRead>(`${BASE}/vendor-conditions/${id}`, {
      method: "PUT",
      body: JSON.stringify(payload),
    }),
};

// ---------------------------------------------------------------------------
// US-ERP-03-04 — Vendor Invoices + 3-way match
// ---------------------------------------------------------------------------

export type VendorInvoiceStatus =
  | "pending"
  | "matched"
  | "tolerance_ok"
  | "blocked"
  | "approved"
  | "paid";

export interface VendorInvoiceRead {
  id: string;
  invoice_number: string;
  vendor_id: string;
  po_id: string;
  gr_id: string | null;
  invoice_date: string;
  total_amount: string;
  currency: string;
  status: VendorInvoiceStatus;
  payment_block: boolean;
  match_details: Record<string, unknown> | null;
  created_at: string;
}

export interface VendorInvoiceCreatePayload {
  invoice_number: string;
  vendor_id: string;
  po_id: string;
  gr_id?: string | null;
  invoice_date: string;
  total_amount: string;
  currency?: string;
}

export interface InvoiceToleranceRead {
  id: string;
  document_type: string;
  vendor_category: string | null;
  tolerance_key: string;
  absolute_limit: string | null;
  pct_limit: string | null;
  currency: string;
  is_active: boolean;
}

export interface VendorInvoiceFilters {
  vendor_id?: string;
  status?: VendorInvoiceStatus;
  limit?: number;
}

export const vendorInvoicesApi = {
  list: (filters: VendorInvoiceFilters = {}): Promise<VendorInvoiceRead[]> =>
    authedFetch<VendorInvoiceRead[]>(
      `${BASE}/invoices${buildQuery(filters as Record<string, string | number | boolean | undefined | null>)}`,
    ),

  create: (payload: VendorInvoiceCreatePayload): Promise<VendorInvoiceRead> =>
    authedFetch<VendorInvoiceRead>(`${BASE}/invoices`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  match: (id: string): Promise<VendorInvoiceRead> =>
    authedFetch<VendorInvoiceRead>(`${BASE}/invoices/${id}/match`, {
      method: "POST",
    }),

  releaseBlock: (id: string, reason: string): Promise<VendorInvoiceRead> =>
    authedFetch<VendorInvoiceRead>(`${BASE}/invoices/${id}/release-block`, {
      method: "POST",
      body: JSON.stringify({ reason }),
    }),
};

export const invoiceTolerancesApi = {
  list: (activeOnly = true): Promise<InvoiceToleranceRead[]> =>
    authedFetch<InvoiceToleranceRead[]>(
      `${BASE}/invoice-tolerances${buildQuery({ active_only: activeOnly })}`,
    ),

  create: (payload: Partial<InvoiceToleranceRead>): Promise<InvoiceToleranceRead> =>
    authedFetch<InvoiceToleranceRead>(`${BASE}/invoice-tolerances`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  update: (id: string, payload: Partial<InvoiceToleranceRead>): Promise<InvoiceToleranceRead> =>
    authedFetch<InvoiceToleranceRead>(`${BASE}/invoice-tolerances/${id}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    }),
};

// ---------------------------------------------------------------------------
// US-ERP-03-05 — Source List + RFQ
// ---------------------------------------------------------------------------

export interface SourceListRead {
  id: string;
  product_sku: string;
  vendor_id: string;
  vendor_name: string | null;
  is_preferred: boolean;
  is_blocked: boolean;
  valid_from: string;
  valid_to: string | null;
  fixed_source: boolean;
  notes: string | null;
}

export type RfqStatus = "draft" | "sent" | "responses_received" | "awarded" | "cancelled";

export interface RfqRead {
  id: string;
  rfq_number: string;
  pr_id: string | null;
  status: RfqStatus;
  deadline: string | null;
  notes: string | null;
  created_at: string;
  created_by: string;
}

export interface RfqResponseRead {
  id: string;
  rfq_id: string;
  vendor_id: string;
  unit_price: string | null;
  currency: string;
  lead_time_days: number | null;
  valid_until: string | null;
  notes: string | null;
  responded_at: string | null;
}

export interface RfqComparisonItem {
  vendor_id: string;
  unit_price: string | null;
  currency: string;
  lead_time_days: number | null;
  score: number | null;
}

export interface RfqComparisonOut {
  rfq_id: string;
  rfq_number: string;
  items: RfqComparisonItem[];
}

export const sourceListApi = {
  list: (params: { product_sku?: string; include_blocked?: boolean } = {}): Promise<SourceListRead[]> =>
    authedFetch<SourceListRead[]>(
      `${BASE}/source-list${buildQuery(params as Record<string, string | number | boolean | undefined | null>)}`,
    ),

  create: (payload: Partial<SourceListRead>): Promise<SourceListRead> =>
    authedFetch<SourceListRead>(`${BASE}/source-list`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  update: (id: string, payload: Partial<SourceListRead>): Promise<SourceListRead> =>
    authedFetch<SourceListRead>(`${BASE}/source-list/${id}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    }),

  delete: (id: string): Promise<void> =>
    authedFetch<void>(`${BASE}/source-list/${id}`, { method: "DELETE" }),
};

export const rfqApi = {
  list: (params: { status?: RfqStatus; limit?: number } = {}): Promise<RfqRead[]> =>
    authedFetch<RfqRead[]>(
      `${BASE}/rfqs${buildQuery(params as Record<string, string | number | boolean | undefined | null>)}`,
    ),

  create: (payload: {
    pr_id?: string | null;
    deadline?: string | null;
    notes?: string | null;
    lines: Array<{ product_sku: string; qty: string; uom?: string }>;
  }): Promise<RfqRead> =>
    authedFetch<RfqRead>(`${BASE}/rfqs`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  addResponse: (rfqId: string, payload: {
    vendor_id: string;
    unit_price?: string | null;
    currency?: string;
    lead_time_days?: number | null;
    valid_until?: string | null;
    notes?: string | null;
  }): Promise<RfqResponseRead> =>
    authedFetch<RfqResponseRead>(`${BASE}/rfqs/${rfqId}/responses`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  comparison: (rfqId: string): Promise<RfqComparisonOut> =>
    authedFetch<RfqComparisonOut>(`${BASE}/rfqs/${rfqId}/comparison`),
};

// ---------------------------------------------------------------------------
// US-ERP-03-06 — KPIs Dashboard
// ---------------------------------------------------------------------------

export type SpendPeriod = "30d" | "90d" | "365d";

export interface ProcurementKpiRead {
  open_pr_count: number;
  open_po_count: number;
  pending_invoice_count: number;
  blocked_invoice_amount: string;
  maverick_spend_pct: string;
  avg_po_lead_time_days: string | null;
  on_time_delivery_pct: string | null;
}

export interface SpendAnalysisRead {
  period_days: number;
  by_vendor: Array<{ vendor_id: string; total_amount: string }>;
  by_product: Array<{ product_sku: string; total_amount: string }>;
}

export const procurementKpiApi = {
  kpis: (): Promise<ProcurementKpiRead> =>
    authedFetch<ProcurementKpiRead>(`${BASE}/kpis`),

  spendAnalysis: (period: SpendPeriod = "30d"): Promise<SpendAnalysisRead> =>
    authedFetch<SpendAnalysisRead>(
      `${BASE}/spend-analysis${buildQuery({ period })}`,
    ),
};

File: mt-pricing-backend/app/schemas/procurement.py
"""Pydantic V2 schemas — EP-ERP-03 (Compras P2P)."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# Purchase Requisition
# ---------------------------------------------------------------------------


class PRCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    product_sku: str | None = None
    qty: Decimal = Field(gt=0)
    uom: str = Field(default="UNIT", max_length=32)
    required_date: date | None = None
    cost_center_id: str | None = Field(default=None, max_length=64)
    suggested_vendor_id: UUID | None = None
    estimated_amount: Decimal | None = Field(default=None, ge=0)
    notes: str | None = None


class PROut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    pr_number: str
    requester_id: UUID
    product_sku: str | None
    qty: Decimal
    uom: str
    required_date: date | None
    cost_center_id: str | None
    suggested_vendor_id: UUID | None
    estimated_amount: Decimal | None
    status: str
    notes: str | None
    created_at: datetime
    updated_at: datetime


class PRSubmit(BaseModel):
    model_config = ConfigDict(extra="forbid")


class PRReject(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    reason: str = Field(min_length=1)


# ---------------------------------------------------------------------------
# Approval Decision
# ---------------------------------------------------------------------------


class ApprovalDecisionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    document_id: UUID
    document_type: str
    approver_id: UUID
    action: str
    reason: str | None
    decided_at: datetime


# ---------------------------------------------------------------------------
# Approval Rule
# ---------------------------------------------------------------------------


class ApprovalRuleCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    document_type: str = Field(default="purchase_requisition", max_length=64)
    min_amount: Decimal = Field(default=Decimal("0"), ge=0)
    max_amount: Decimal | None = Field(default=None, ge=0)
    category_id: str | None = Field(default=None, max_length=64)
    approver_role: str | None = Field(default=None, max_length=32)
    approver_user_id: UUID | None = None
    timeout_hours: int = Field(default=48, ge=0)
    priority: int = Field(default=0, ge=0)
    is_active: bool = True


class ApprovalRuleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    document_type: str
    min_amount: Decimal
    max_amount: Decimal | None
    category_id: str | None
    approver_role: str | None
    approver_user_id: UUID | None
    timeout_hours: int
    priority: int
    is_active: bool
    created_at: datetime


class ApprovalRuleUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    min_amount: Decimal | None = Field(default=None, ge=0)
    max_amount: Decimal | None = Field(default=None, ge=0)
    category_id: str | None = None
    approver_role: str | None = Field(default=None, max_length=32)
    approver_user_id: UUID | None = None
    timeout_hours: int | None = Field(default=None, ge=0)
    priority: int | None = Field(default=None, ge=0)
    is_active: bool | None = None


# ---------------------------------------------------------------------------
# Vendor Product Condition (PIR)
# ---------------------------------------------------------------------------


class VendorConditionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    vendor_id: str = Field(min_length=1, max_length=64)
    product_sku: str = Field(min_length=1)
    price: Decimal = Field(ge=0)
    uom: str = Field(default="UNIT", max_length=32)
    moq: int = Field(default=1, ge=1)
    lead_time_days: int | None = Field(default=None, ge=0)
    valid_from: date | None = None
    valid_to: date | None = None
    currency: str = Field(default="AED", min_length=3, max_length=3)
    is_active: bool = True


class VendorConditionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    vendor_id: str
    product_sku: str
    price: Decimal
    uom: str
    moq: int
    lead_time_days: int | None
    valid_from: date
    valid_to: date | None
    currency: str
    is_active: bool
    created_at: datetime


class VendorConditionUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    price: Decimal | None = Field(default=None, ge=0)
    uom: str | None = Field(default=None, max_length=32)
    moq: int | None = Field(default=None, ge=1)
    lead_time_days: int | None = Field(default=None, ge=0)
    valid_to: date | None = None
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    is_active: bool | None = None


# ---------------------------------------------------------------------------
# Vendor Invoices (US-ERP-03-04 — 3-way match)
# ---------------------------------------------------------------------------


class VendorInvoiceCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    invoice_number: str = Field(min_length=1, max_length=128)
    vendor_id: str = Field(min_length=1, max_length=128)
    po_id: UUID
    gr_id: UUID | None = None
    invoice_date: date
    total_amount: Decimal = Field(ge=0)
    currency: str = Field(default="AED", min_length=3, max_length=3)


class VendorInvoiceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    invoice_number: str
    vendor_id: str
    po_id: UUID
    gr_id: UUID | None
    invoice_date: date
    total_amount: Decimal
    currency: str
    status: str
    payment_block: bool
    match_details: dict | None
    created_at: datetime


class InvoiceReleaseBlock(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    reason: str = Field(min_length=1)


class InvoiceToleranceCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    document_type: str = Field(default="vendor_invoice", max_length=64)
    vendor_category: str | None = None
    tolerance_key: str = Field(min_length=1, max_length=128)
    absolute_limit: Decimal | None = Field(default=None, ge=0)
    pct_limit: Decimal | None = Field(default=None, ge=0)
    currency: str = Field(default="AED", min_length=3, max_length=3)
    is_active: bool = True


class InvoiceToleranceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    document_type: str
    vendor_category: str | None
    tolerance_key: str
    absolute_limit: Decimal | None
    pct_limit: Decimal | None
    currency: str
    is_active: bool


class InvoiceToleranceUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    absolute_limit: Decimal | None = Field(default=None, ge=0)
    pct_limit: Decimal | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    is_active: bool | None = None


# ---------------------------------------------------------------------------
# Source List (US-ERP-03-05)
# ---------------------------------------------------------------------------


class SourceListCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    product_sku: str = Field(min_length=1)
    vendor_id: str = Field(min_length=1, max_length=128)
    vendor_name: str | None = None
    is_preferred: bool = False
    is_blocked: bool = False
    valid_from: date | None = None
    valid_to: date | None = None
    fixed_source: bool = False
    notes: str | None = None


class SourceListOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    product_sku: str
    vendor_id: str
    vendor_name: str | None
    is_preferred: bool
    is_blocked: bool
    valid_from: date
    valid_to: date | None
    fixed_source: bool
    notes: str | None


class SourceListUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    vendor_name: str | None = None
    is_preferred: bool | None = None
    is_blocked: bool | None = None
    valid_to: date | None = None
    fixed_source: bool | None = None
    notes: str | None = None


# ---------------------------------------------------------------------------
# RFQ (US-ERP-03-05)
# ---------------------------------------------------------------------------


class RfqLineCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    product_sku: str = Field(min_length=1)
    qty: Decimal = Field(gt=0)
    uom: str = Field(default="UNIT", max_length=32)


class RfqLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    rfq_id: UUID
    product_sku: str
    qty: Decimal
    uom: str


class RfqCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    pr_id: UUID | None = None
    deadline: date | None = None
    notes: str | None = None
    lines: list[RfqLineCreate] = Field(min_length=1)


class RfqOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    rfq_number: str
    pr_id: UUID | None
    status: str
    deadline: date | None
    notes: str | None
    created_at: datetime
    created_by: UUID


class RfqResponseCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    vendor_id: str = Field(min_length=1, max_length=128)
    unit_price: Decimal | None = Field(default=None, ge=0)
    currency: str = Field(default="AED", min_length=3, max_length=3)
    lead_time_days: int | None = Field(default=None, ge=0)
    valid_until: date | None = None
    notes: str | None = None


class RfqResponseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: UUID
    rfq_id: UUID
    vendor_id: str
    unit_price: Decimal | None
    currency: str
    lead_time_days: int | None
    valid_until: date | None
    notes: str | None
    responded_at: datetime | None


class RfqComparisonItem(BaseModel):
    """Un renglón en la tabla comparativa de respuestas RFQ."""

    vendor_id: str
    unit_price: Decimal | None
    currency: str
    lead_time_days: int | None
    score: float | None
    """Score compuesto: 0.6*(1/precio_norm) + 0.4*(1/lead_time_norm). Mayor es mejor."""


class RfqComparisonOut(BaseModel):
    rfq_id: UUID
    rfq_number: str
    items: list[RfqComparisonItem]


# ---------------------------------------------------------------------------
# KPIs Dashboard (US-ERP-03-06)
# ---------------------------------------------------------------------------


class ProcurementKpiOut(BaseModel):
    """KPIs consolidados del módulo de compras."""

    open_pr_count: int
    open_po_count: int
    pending_invoice_count: int
    blocked_invoice_amount: Decimal
    maverick_spend_pct: Decimal
    avg_po_lead_time_days: Decimal | None
    on_time_delivery_pct: Decimal | None


class SpendByVendor(BaseModel):
    vendor_id: str
    total_amount: Decimal


class SpendByProduct(BaseModel):
    product_sku: str
    total_amount: Decimal


class SpendAnalysisOut(BaseModel):
    period_days: int
    by_vendor: list[SpendByVendor]
    by_product: list[SpendByProduct]
